apis/v1alpha2/gateway_types.go

```diff
@@ -157,10 +158,22 @@ type Listener struct {
 	// field is ignored for protocols that don't require hostname based
 	// matching.
 	//
+	// Implementations MUST apply Hostname matching appropriately for each of
+	// the following protocols:
+	//
+	// * TLS: The Hostname match MUST be applied to the
+	//   [SNI](https://tools.ietf.org/html/rfc6066#section-3) server name
+	//   offered by the client.
+	// * HTTP: The Hostname match MUST be applied to the host portion of the
+	//   [effective request URI](https://tools.ietf.org/html/rfc7230#section-5.5)
+	//   or the [:authority pseudo-header](https://tools.ietf.org/html/rfc7540#section-8.1.2.3)
+	// * HTTPS: The Hostname match MUST be applied at both the TLS and HTTP
+	//   protocol layers.
```

Comment:
@youngnick Is this sufficient to close the domain fronting issue (#72) as well?

Reply:
Although this isn't entirely new content, now that I'm seeing it here, I think this wording may be a bit too strong. I'm not sure that every implementation _can_ apply a match at both layers.

Reply:
I think the direction is good, but maybe we need to specify this in terms of API objects, like:
```
- TLS: the Listener hostname must match the SNI.
- HTTP: The HTTPRoute hostname must match the Host header of the request.
- HTTPS: The Listener Hostname and the HTTPRoute hostname must agree with each other, and also match request properties as above.
```

Reply:
Because the important part is that we expect that Listeners and HTTPRoutes that do not match these rules should not end up attached.

Reply:
I don't think we should mix up the route type with the listener protocol here. While there's likely a ~99% overlap here, if we tie route types too closely here, we prevent or at least complicate extended route types. I've copied most of your suggestions from above, but removed any references to Routes - the following paragraph already covers that intersection.